graphium/trainer/losses.py

```diff
@@ -52,22 +52,42 @@ def __init__(
         self.alpha = alpha
         self.softmax = torch.nn.Softmax(dim=1)
 
-    def forward(self, input: Tensor, target: Tensor) -> Tensor:
+    def forward(self, input: Tensor, target: Tensor, nan_targets: Tensor = None) -> Tensor:
         """
         Parameters:
             input: (batch_size x n_classes) tensor of logits predicted for each bracket.
             target: (batch_size) or (batch_size, 1) tensor of target brackets in {0, 1, ..., self.n_brackets}.
         """
 
         target = target.flatten()
+
+        # set input and target with nans to 0s for regression loss
+        if nan_targets is not None:
+            input[nan_targets] = 0.0
+            target[nan_targets] = 0.0
         # regression loss needs normalized logits to probability as input to do inner product with self.brackets
         # we apply softmax on the raw logits first
         softmax_input = self.softmax(input)
+        # the softmax of a tensor of 0s would not be 0s anymore, so need to apply nan_targets here again to filter out
+        if nan_targets is not None:
+            softmax_input[nan_targets] = 0.0
         # [batch_size, n_classes] * [n_classes] ([0, 1, 2...n_brakets-1]) -> [batch_size]
         regression_input = torch.inner(softmax_input, self.brackets.to(input.device))
         regression_loss = self.regression_loss(regression_input, target.float(), reduction=self.reduction)
+        # regression_loss needs some scaling by total_targets/num_real_targets
+        if nan_targets is not None:
+            num_real_targets = (~nan_targets).sum()
+            factor1 = torch.where(num_real_targets > 0, 1, 0)
+            factor2 = torch.where(num_real_targets > 0, 0, 1)
+            regression_loss = factor1 * regression_loss * nan_targets.numel() / (num_real_targets + factor2)
 
+            # set input and target with nans to -1000s for ce loss
+            input[nan_targets] = -1000
+            target[nan_targets] = -1000
         # cross_entropy loss needs raw logits as input
-        ce_loss = F.cross_entropy(input, target.long(), weight=self.weight, reduction=self.reduction)
+        # ce_loss does not need scaling as it already ignores -1000 masked nan values
+        ce_loss = F.cross_entropy(
+            input, target.long(), weight=self.weight, ignore_index=-1000, reduction=self.reduction
+        )
```

Comment:
Found a bug here: the nans are set to 0s in `ipu_losses.py` and here `softmax` is applied inside regression and ce losses, while the `softmax` of 0s are not 0s. Previous approach would lead to wrong estimation of hybridceloss. 
1, for regression_loss, it is easy to filter that out by nan_targets
2, for ce_loss softmax is integrated with the `F.cross_entropy`, so used -1000 to mark the nans and ask the ce loss to ignore -1000.
3, this means ce_loss does not need nan_target scaling any more, so we do scaling only on regression_loss

Reply:
I think we need the `F.cross_entroy` as it has labels of [0, 1, 2, 3, 4]? Is `BCEwithLogitsLossIPU` just used for binary cross entropy? For regression loss, I think indeed we can pass the logits after softmax to the `MSELossIPU`. I think the point here is that, for regression loss, input[nan_target] = 0.0 and softmax_input[nan_target] = 0.0 need to be applied, for ce loss, `F.cross_entropy` includes a softmax, we can't set nans to 0s and hope that the ce loss would ignore that as softmax of 0s is not 0s.